ultravox/data/datasets.py

```diff
@@ -1060,43 +1068,73 @@ def create_dataset(name: str, args: VoiceDatasetArgs) -> data.IterableDataset:
         Args:
-            datasets: a list of IterableDataset objects
-            repeat: whether to repeat the datasets indefinitely.
-                This matters most when the datasets have different lengths.
-                Let's say you have two datasets, A and B which have 5 and 3 samples respectively.
-
-                `repeat=False`: [A0, B0, A1, B1, A2, B2, A3, A4]
-                `repeat=True` : [A0, B0, A1, B1, A2, B2, A3, B0, A4, B1, A0, ...]
-
-                NOTE: with `repeat=True`, `__iter__` never stops.
+            datasets: A list of data.IterableDataset objects.
+            stop_strategy: Strategy for stopping iteration. Choose from 'first_exhausted', 'last_exhausted', 'never_stop'.
+            seed: Optional seed for reproducibility.
+            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         super().__init__()
         self._datasets = datasets
-        self._repeat = repeat
+        self._rng = np.random.default_rng(seed)
+        self._static = static
+
+        if stop_strategy not in ["first_exhausted", "last_exhausted", "never_stop"]:
+            raise ValueError(
+                "Invalid stop_strategy. Choose from 'first_exhausted', 'last_exhausted', 'never_stop'."
+            )
+        self._stop_strategy = stop_strategy
+
+        weights = [getattr(ds, "get_weight", lambda: 1)() for ds in datasets]
+        total_weight = sum(weights)
+        self._normalized_probs = [w / total_weight for w in weights]
 
     def __iter__(self):
+        # If no datasets are provided, return an empty iterator
+        if not self._datasets:
+            return
+
         iters = [iter(ds) for ds in self._datasets]
-        iter_index = 0
+        exhausted = [False] * len(iters)
+
+        if self._static:
+            static_iter = cycle(range(len(self._datasets)))
+
+        while True:
+            if self._static:
+                iter_index = next(static_iter)
+            else:
+                iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
 
-        while len(iters):
-            it = iters[iter_index]
             try:
-                val = next(it)
-                iter_index = (iter_index + 1) % len(iters)
-                yield val
+                yield next(iters[iter_index])
             except StopIteration:
-                if not self._repeat:
-                    iters.pop(iter_index)
-                    iter_index %= max(1, len(iters))
-                else:
-                    iters[iter_index] = iter(self._datasets[iter_index])
+                # Reconstruct the iterator
+                iters[iter_index] = iter(self._datasets[iter_index])
+                exhausted[iter_index] = True
+                if self._static:
+                    static_iter = cycle(
+                        list(range(iter_index + 1, len(self._datasets)))
+                        + list(range(iter_index + 1))
+                    )
```

Comment:
IIUC if dataset A is exhausted then dataset B, the dataset A gets added back into the mix.

It seems like a mix of the `repeat=False` and `repeat=True` behaviour. Which one did you want?

Reply:
When we do weighted sampling, we don't have to have `repeat=False`